runtime/drivers/motherduck/duckdb.go

```diff
@@ -0,0 +1,223 @@
+		"LOAD 'httpfs'",
+		"SET max_expression_depth TO 250",
+	}
+
+	// DuckDB extensions need to be loaded separately on each connection, but the built-in connection pool in database/sql doesn't enable that.
+	// So we use go-duckdb's custom connector to pass a callback that it invokes for each new connection.
+	// nolint:staticcheck // TODO: remove when go-duckdb implements the driver.ExecerContext interface
+	connector, err := duckdb.NewConnector(cfg.DSN, func(execer driver.ExecerContext) error {
+		for _, qry := range bootQueries {
+			_, err = execer.ExecContext(context.Background(), qry, nil)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		// Adding check if using old database files with upgraded version
+		if strings.Contains(err.Error(), "Trying to read a database file with version number") {
+			return nil, fmt.Errorf("database file %q was created with an older, incompatible version of Rill (please remove it and try again)", cfg.DSN)
+		}
+
+		if strings.Contains(err.Error(), "Could not set lock on file") {
+			return nil, fmt.Errorf("failed to open database (is Rill already running?): %w", err)
+		}
+
+		return nil, err
+	}
+
+	// This driver may issue both OLAP and "meta" queries (like catalog info) against DuckDB.
+	// Meta queries are usually fast, but OLAP queries may take a long time. To enable predictable parallel performance,
+	// we gate queries with semaphores that limits the number of concurrent queries of each type.
+	// The metaSem allows 1 query at a time and the olapSem allows cfg.PoolSize-1 queries at a time.
+	//
+	// When cfg.PoolSize is 1, we set olapSem to still allow 1 query at a time.
+	// This creates contention for the same connection in database/sql's pool, but its locks will handle that.
+
+	sqlDB := otelsql.OpenDB(connector)
+	db := sqlx.NewDb(sqlDB, "duckdb")
+	db.SetMaxOpenConns(cfg.PoolSize)
+
+	// We want to use all except one connection for OLAP queries.
+	olapSemSize := cfg.PoolSize - 1
+	if olapSemSize < 1 {
+		olapSemSize = 1
+	}
+
+	c := &connection{
+		db:      db,
+		metaSem: semaphore.NewWeighted(1),
+		olapSem: priorityqueue.NewSemaphore(olapSemSize),
+		logger:  logger,
+		config:  cfg,
+	}
+
+	// Return nice error for old macOS versions
+	conn, err := c.db.Connx(context.Background())
+	if err != nil && strings.Contains(err.Error(), "Symbol not found") {
+		fmt.Printf("Your version of macOS is not supported. Please upgrade to the latest major release of macOS. See this link for details: https://support.apple.com/en-in/macos/upgrade")
+		os.Exit(1)
+	} else if err == nil {
+		conn.Close()
+	} else {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (d Driver) Drop(dsn string, logger *zap.Logger) error {
+	cfg, err := newConfig(dsn)
+	if err != nil {
+		return err
+	}
+
+	if cfg.DBFilePath != "" {
+		err = os.Remove(cfg.DBFilePath)
+		if err != nil {
+			return err
+		}
+		// Hacky approach to remove the wal file
+		_ = os.Remove(cfg.DBFilePath + ".wal")
+	}
```

Comment:
Will duckdb still create `.db` files locally when using motherduck? Or could it just always pass `:memory:`?

Reply:
No it doesn't. Also realised that we need to drop the instance specific database. Added a hacky impl to do that.